README.md
# 3N1GM4
Multipurpose Discord bot, developed by: `Anonymous©#7296`.

## Assumptions
I've always wanted to code my own bot. I've never been sure what kind of bot exactly I want to code.
In February 2020 I've started reading about Discord.py library and I've coded mu first Discord bot.
It was very simple bot, it was presenting some useful data for economic browser game.
Then I wanted to make something bigger and started to code `3N1GM4` bot.

## Functions:
All functionalities should get data from SQL databases.

All modules can be disabled per guild. "Manage guild" permission will be required.

###### User profile
- [x] avatar URL

###### Role playing
I think cyberpunk will be the best lore with cryptocurrency as virtual money.
- [ ] quests
- [ ] virtual wallet
- [ ] daily bonus

###### Guild stats
- [ ] users
- [ ] bots
- [ ] online
- [ ] bans

###### Support
- [ ] creating ticket with reaction
- [ ] creating ticket by sending a message 

###### Funny
- [ ] joke
- [ ] slap someone or random
- [ ] psycho test (pseudorandom)
- [ ] IQ test (pseudorandom)
    - both tests will be based on user ID and discriminator
- [ ] random name generator
- [ ] random pick
- [ ] random color
- [ ] 8ball

###### Images
[How to source](https://stackoverflow.com/questions/57043797/discord-py-getting-random-imgur-images)
- [ ] random meme from Reddit
- [ ] random (or not) image from Imigur

###### Reputation
- [ ] giving `rep`'s to someone (once a day)
- [ ] best users in guild

###### Commands settings
- [ ] enabling/disabling commands per guild

###### Other utilities
- [ ] reminder
- [ ] notes
- [ ] data caching
- [ ] debug option to some commands
- [ ] how to do something
- [ ] curiosities
- [ ] giveaways

###### Admin purpose
- [x] ban
- [x] kick
- [ ] censoring bad words (different languages with community help)



docs/CHANGELOG.md
# Enigma Bot changelog

## [Unreleased]

## Released

### [v0.8] - 11.02.2021
#### Added
- Logging when bot is online and ready.
- Logging Discord connection states.
#### Fixed
- Logging after leaving a server.
- `spotify` command now works with links which include `&` or `?`.
#### Updated
- [`setup.py`] now has all packages
#### Changed
- Python version in [`setup.py`] is now read from [`runtime.txt`]
- There's author's full name in [`setup.py`].
- Name in [`setup.py`] now looks better.
- `info` command embed.
- `spotify` command help.
- `spotify` command cooldown, from 1 per 30s to 3 per 60s.

### v0.7-hotfix - 09.02.2021
#### Fixed
- `spotify` command was disabled in production.

### [v0.7] - 08.02.2021
#### Added
- `spotify` command.
#### Fixed
- `ping` commands was showing negative values.
- `giveaway` command embed.
#### Changed
- `suggest` command aliases and description.
- Settings are now more Python-friendly.

### [v0.6] - 04.02.2021
#### Added
- `unban` command.
- `servers` command.
- `whois` command.
- Logging guilds count.
- Logging amount of cogs loaded.
- Logging connection to Discord.
- Checking connection with database.
#### Fixed
- `randomnumber` command was disabled.
- `host` command without na argument no more raises an error.
#### Changed
- Command errors are now handled globally.
- `ping` command sends both ping and latency.
- `help` command do not sends classic representation if it's same as usage.
- `prune` command's error description slightly changed to be more precise.
- Logging is now more clear.
- SQL-Alchemy logging level set to WARNING.
- Version in settings is now a function.
- Normalized get function in `settings.py`.
#### Removed
- `error` command.

### [v0.5] - 03.02.2021
#### Added
- `suggest` command.
- `prune` command.
#### Changed
- Embeds have now fixed colors, based on message type.
- `error` cooldown increased from 30 seconds to 60.

### [v0.4] - 25.10.2020
#### Added
- `randomnumber` command.
- `choice` command.
- `coin` command.
#### Fixed
- Communication with database is now more safe.

### v0.3-hotfix
#### Fixed
- `help` command was disabled.

### [v0.3] - 21.10.2020
#### Added
- `invite` command.
- Error when giveaway result is too big is handled.
- `avatar`'s embed description simplified.
- `daily` command.
- `info` command.
- Custom `help` command.
#### Changed
- Using `giveaway` command requires "manage guild" permission.
- Increased `meme`'s cooldown from 4s to 6s.
- `giveaway`'s help is now more precise.
- `giveaway`'s name has to be 30 characters or less.
- `giveaway`'s quantity has to be 25 or less.
- Giveaway's preview message and final message have now fields instead of big descriptions.
- `user` parameter is now optional in `profile` command.
- Profile's now are on field instead of big description.
- Optional 3rd argument in `giveaway` is now asked for.
- Final `giveaway` massage is prettier.
- Cooldowns' messages no more include time, because it was incorrect.
- All commands are now enabled only in development or production.
#### Fixed
- First `daily` is not creating profile **and** adds cash.
- Value 0 is now possible in `manage` command.
- Using `daily` 2 and more times now shows correct message.

### [v0.2] - 16.10.2020
#### Added
- `meme` command.
- Choosing winners from giveaway.
- Optional argument how to group winners from giveaway.
- Cosmetics to giveaways' messages.
- Cooldown for `giveaway` and `iq`.
#### Changed
- Cooldown handler text is same everywhere.
- If `iq`'s command argument is Enigma bot itself it sends score 200.
- If `iq`'s command argument is any other bot it sends score 0.

### [v0.1] - 16.10.2020
#### Added
- Banning command.
- Kicking command.
- Checking bot's latency command.
- Giveaways command.
- Getting users' avatars command.
- Managing users' data command.
- Users' profiles command.
- Communication with the database using SQLAlchemy.


[`setup.py`]: https://github.com/AnonymousX86/Enigma-Bot/blob/master/setup.py
[`runtime.txt`]: https://github.com/AnonymousX86/Enigma-Bot/blob/master/runtime.txt
[Unreleased]: https://github.com/AnonymousX86/Enigma-Bot/compare/deploy...master
[v0.8]: https://github.com/AnonymousX86/Enigma-Bot/releases/tag/v0.8
[v0.7]: https://github.com/AnonymousX86/Enigma-Bot/releases/tag/v0.7
[v0.6]: https://github.com/AnonymousX86/Enigma-Bot/releases/tag/v0.6
[v0.5]: https://github.com/AnonymousX86/Enigma-Bot/releases/tag/v0.5
[v0.4]: https://github.com/AnonymousX86/Enigma-Bot/releases/tag/v0.4
[v0.3]: https://github.com/AnonymousX86/Enigma-Bot/releases/tag/v0.3
[v0.2]: https://github.com/AnonymousX86/Enigma-Bot/releases/tag/v0.2
[v0.1]: https://github.com/AnonymousX86/Enigma-Bot/releases/tag/v0.1



enigma/cogs/basics.py
# -*- coding: utf-8 -*-
from datetime import timedelta, datetime

from discord import TextChannel
from discord.ext.commands import command, Cog, MissingPermissions, Context, cooldown, BucketType, \
    CommandError, Command

from enigma.emebds.core import InfoEmbed, SuccessEmbed, ErrorEmbed
from enigma.emebds.misc import SuggestionEmbed
from enigma.settings import in_production, suggestions_channel_id


class Basics(Cog):
    def __init__(self, bot):
        self.bot = bot

    @command(
        name='help',
        brief='Shows this message',
        usage='[category|command]',
        enabled=in_production()
    )
    async def help(self, ctx: Context, arg: str = None):
        help_em = InfoEmbed(
            author=ctx.author,
            title=':grey_question: Help'
        )
        help_category = 'You can also type `>help <category>` for more info on a category.'
        help_command = 'Type `>help <command>` for more info on a command.'
        if arg is None:
            help_em.description = f'{help_category}\n{help_command}'
            for cog in self.bot.cogs:
                help_em.add_field(
                    name=str(cog),
                    value=', '.join(map(lambda x: f'`{x.name}`', self.bot.cogs[cog].get_commands())),
                    inline=False
                )
        else:
            cmd: Command = self.bot.get_command(arg)
            if not cmd:
                cog: Cog = self.bot.get_cog(arg.capitalize())
                if not cog:
                    help_em.description = f'No command or category called `{arg}` found.'
                # Cog found
                else:
                    help_em.description = help_command
                    help_em.add_field(
                        name='Category',
                        value=f'```\n{cog.qualified_name}\n```',
                        inline=False
                    )
                    for cmd in cog.get_commands():
                        try:
                            if await cmd.can_run(ctx):
                                help_em.add_field(
                                    name=f'`{cmd}`',
                                    value=cmd.brief or '\u200b'
                                )
                        except CommandError:
                            pass
            # Command found
            else:
                help_em.description = help_category
                help_em.add_field(
                    name='Command',
                    value=f'```\n{cmd.name}\n```',
                    inline=False
                )
                if cmd.description:
                    help_em.add_field(
                        name='Description',
                        value=cmd.description,
                        inline=False
                    )
                if cmd.aliases:
                    help_em.add_field(
                        name='Aliases',
                        value=', '.join(map(lambda x: f'`{x}`', cmd.aliases)),
                        inline=False
                    )
                if cmd.usage:
                    help_em.add_field(
                        name='Usage',
                        value=f'```\n>{cmd.name} {cmd.usage}\n```',
                        inline=False
                    )
                if cmd.help:
                    help_em.add_field(
                        name='Additional help',
                        value=cmd.help,
                        inline=False
                    )
                if cmd.aliases or cmd.usage:
                    base = cmd.name if not cmd.aliases else f'[{"|".join([cmd.name, *cmd.aliases])}]'
                    tail = cmd.usage or ''
                    if f'{cmd.name} {cmd.usage}' != f'{base} {tail}':
                        help_em.add_field(
                            name='Classic representation',
                            value=f'```\n>{base} {tail}\n```'
                        )

        await ctx.send(embed=help_em)

    @command(
        name='ping',
        brief='Checks bot latency',
        description='Counts time difference between command execution time and bot\'s response',
        enabled=in_production()
    )
    async def ping(self, ctx: Context):
        elapsed_time: timedelta = datetime.utcnow() - ctx.message.created_at
        m, s = divmod(elapsed_time.total_seconds(), 60)
        ping = int(round((m * 60 + s) * 1000))
        await ctx.send(embed=SuccessEmbed(
            author=ctx.author,
            title=':ping_pong: Pong!'
        ).add_field(
            name='Latency',
            value=f'{round(self.bot.latency * 1000)}ms'
        ).add_field(
            name='Ping',
            value=f'{ping}ms'
        ))

    @command(
        name='invite',
        brief='Sends bot\'s invite link',
        enabled=in_production()
    )
    async def invite(self, ctx: Context):
        await ctx.send(embed=InfoEmbed(
            author=ctx.author,
            title=':mailbox_with_mail: Check at top.gg',
            description='**\u00bb [Click me!](https://top.gg/bot/678357487560425555) \u00ab**'
        ).add_field(
            name='Important info',
            value='Remember, that you need **manage users** permission to add me to the server.'
        ))

    @command(
        name='info',
        brief='Sends short bot\'s info',
        description='Sends info about bot itself and its author. This is NOT help command.',
        aliases=['about', 'github', 'code'],
        enabled=in_production()
    )
    async def info(self, ctx: Context):
        await ctx.send(embed=InfoEmbed(
            author=ctx.author,
            title=':desktop: Source code',
            description='This bot is made on open source, GNU GPL v3.0 license.'
        ).add_field(
            name='Links',
            value='\u00b7 [GitHub homepage](https://github.com/AnonymousX86/Enigma-Bot)\n'
                  '\u00b7 [Changelog](https://github.com/AnonymousX86/Enigma-Bot/'
                  'blob/master/docs/CHANGELOG.md#enigma-bot-changelog)\n',
            inline=False
        ).add_field(
            name='Support',
            value='If you\'d like to help to develop this bot please check GitHub. (Link above)\n'
                  'And if **you** need support please check'
                  ' [official support server](https://discord.gg/SRdmrPpf2z "3N1GMA Support Server").',
            inline=False
        ).add_field(
            name='More about bot',
            value=f'Version: `{self.bot.version}`.\n'
                  f'Author: `{str(self.bot.get_user(self.bot.owner_id))}`.',
            inline=False
        ))

    @cooldown(1, 120, BucketType.user)
    @command(
        name='suggest',
        brief='Suggest a change',
        description='Ask dev(s) for a functionality or give a feedback.',
        help='Keep your message between 25 and 120 characters.',
        usage='<message>',
        aliases=['change', 'feedback'],
        enabled=in_production()
    )
    async def change(self, ctx: Context, *, message: str = None):
        if not message:
            await ctx.send(embed=ErrorEmbed(
                author=ctx.author,
                title=':x: Whoops!',
                description='You forget to add a message.'
            ))
        elif len(message) < 25:
            await ctx.send(embed=ErrorEmbed(
                author=ctx.author,
                title=':x: Whoops!',
                description='Suggestion is too short. Write more, please.'
            ))
        elif len(message) > 120:
            await ctx.send(embed=ErrorEmbed(
                author=ctx.author,
                title=':x: Whoops!',
                description='Suggestion is too long. Write less, please.'
            ))
        else:
            channel: TextChannel = self.bot.get_channel(suggestions_channel_id())
            msg = await channel.send(embed=SuggestionEmbed(
                author=ctx.author,
                message=message
            ))
            await ctx.send(embed=SuccessEmbed(
                author=ctx.author,
                title=':thumbsup: Thanks for suggestion!',
                description=f'Check your suggestion here: {channel.mention} or click'
                            f' [here]({msg.jump_url} "Direct message link").'
            ))
            await msg.add_reaction(emoji='👍')
            await msg.add_reaction(emoji='👎')

    @cooldown(1, 60, BucketType.guild)
    @command(
        name='servers',
        brief='List of guilds bot is in',
        aliases=['guilds', 'serverlist', 'guildlist', 'slist'],
        enabled=not in_production(),
        hidden=True
    )
    async def servers(self, ctx: Context):
        if ctx.author.id != self.bot.owner_id:
            raise MissingPermissions(missing_perms=['bot_owner'])
        em = InfoEmbed(
            author=ctx.author,
            title=':passport_control: Bot\'s servers list',
            description='Last 10 (or less) servers'
        )
        for g, num in zip(self.bot.guilds[:10], 'one,two,three,four,five,six,seven,eight,nine,one::zero'.split(',')):
            em.add_field(
                name=f':{num}:  -  {g.name}',
                value=f'```py\n'
                      f'Created at:    {str(g.created_at)[:10]}\n'
                      f'Joined at:     {str(g.me.joined_at)[:10]}\n'
                      f'Member count:  {g.member_count}\n'
                      f'Owner:         {str(g.owner)}\n'
                      f'Premium tier:  {g.premium_tier}\n'
                      f'Features:      {", ".join(g.features) or "-"}\n'
                      f'```',
                inline=False
            )
        await ctx.send(embed=em)

    # TODO - Ticket system


def setup(bot):
    bot.add_cog(Basics(bot))
